source/common/stats/thread_local_store.cc

```diff
@@ -199,6 +202,15 @@ void ThreadLocalStoreImpl::initializeThreading(Event::Dispatcher& main_thread_di
 void ThreadLocalStoreImpl::shutdownThreading() {
   // This will block both future cache fills as well as cache flushes.
   shutting_down_ = true;
+
+  // We can't call runOnAllThreads here as global threading has already been shutdown.
```

Comment:
WDYT of adding an API `ThreadLocal::Instance::isShutdown()` and asserting that here?

I actually think this is *not* true in the current unit tests, where I see:
```
  ~ThreadLocalRealThreadsTestBase() override {
    {
      BlockingBarrier blocking_barrier(1);
      main_dispatcher_->post(blocking_barrier.run([this]() {
        store_->shutdownThreading();
        tls_->shutdownGlobalThreading();
        tls_->shutdownThread();
      }));
    }
```
But you can re-order (I just checked in a clean client) to:
```
        tls_->shutdownGlobalThreading();
        store_->shutdownThreading();
        tls_->shutdownThread();
```
so that my suggested invariant holds.

I think if there's some usage of this library that has the order as it currently sits in the unit test destrructor above, I'd be worried about a race if global threading has *not* been shut down, thus my suggestion to assert it.

Reply:
I removed all the calls to shutdownGlobalThreading I had added.
However, I did leave this re-ordering:
` 
  tls_->shutdownGlobalThreading();
  store_->shutdownThreading();
  tls_->shutdownThread();
` 

This is still our intent, right?

WRT adding `ThreadLocal::Instance::isShutdown()`, just wondering if that should be tied to the main thread and call it `ThreadLocal::Instance::isGlobalThreadingShutdown()`. I am open to either.

Reply:
Here's a gist that adds isShutdown: https://gist.github.com/jmarantz/46059fac142b0d089174a23127b6e1b1

Reply:
Appreciate the assist. I may have been over complicating that one, will get that applied and building today.

Reply:
It needs to be cleaned up -- e.g. the interface function should be pure, and it needs doxygen doc. There may be other implementations of that needing to be filled in.